### PR TITLE
Use different symbol names for caml_do_local_roots on bytecode and native

### PR DESCRIPTION
**Pull request by stedolan**
In #8807, a custom Makefile that manually linked the threads library and the runtime built a binary that failed with a mysterious segfault. The problem turned out to be that the bytecode version of systhreads was being linked against the native runtime.

This should have been a link error, but since the two runtimes use the same name for the `caml_do_local_roots` function there was no error. Since the signatures are different, it segfaulted at runtime.

This PR contains a tiny patch to ensure that bytecode and native use different symbol names for this function.

### CODE REVIEW

**Comment by dra27** [comment]
Hmm, this is a strange vindication of both ocamltest's and Travis's pernicketyness! This fails with `ocamlc.opt` unless you've disabled shared library support precisely because linking attempts to load all `.so` files to check that they link. This of course fails for `ocamlc.opt` because it provides `caml_do_local_roots_nat`.

**Comment by stedolan** [comment]
Thanks for the explanation!

I hadn't realised that ocamlc tries to dlopen any libraries you've asked it to link against. Is this really wise? Shared objects can do weird things at startup, and loading them all into the compiler sounds suspicious. (In particular, the native and bytecode versions of the threads library are not binary compatible, so here we're loading the wrong one)

**Comment by dra27** [comment]
An alternative, on platforms which support it, would be to use libbfd to get the primitives list, which would be superior (but a larger change).

A possibly simpler workaround, given that this comes from a real-life error, might be to link an extra primitive into `ocamlc.opt` when shared library support is enabled?!

**Comment by dra27** [comment]
A sort-of-implementation of the halfway house. Assuming it does work, it deals with the original problem - you can no longer link the bytecode stubs library for systhreads against the native runtime but of course the converse is true if you use `ocamlc.opt` and try to link against threadsnat. That's a somewhat harder and weirder mistake to make than not knowing about `-lthreadsnat` vs `-lthreads` so maybe there's some value in that commit?

File: runtime/caml/roots.h
Comment (dra27) [inline runtime/caml/roots.h]: If this ends up going in some form, I think it might be clearer to rename the functions in `roots_byt.c` and `roots_nat.c` to the `_byt` and `_nat` versions and put the `#define` after the `CAMLextern` using the actual name!

**Comment by dra27** [comment]
While reading up on another PR, I came across https://github.com/ocaml/ocaml/issues/6845 which is relevant to the discussion here!

**Comment by stedolan** [comment]
(I'll hold off on updating this until #9551 is resolved - if that gets merged, the ocamlc.opt issue might go away)

File: runtime/caml/roots.h
Comment (stedolan) [reply runtime/caml/roots.h]: Good idea, done

**Comment by stedolan** [comment]
Now that #9551 is merged, the original implementation seems to work fine - ocamlc.opt no longer tries to dynamically load bytecode shared libraries.

**Comment by xavierleroy** [comment]
> Now that #9551 is merged, the original implementation seems to work fine - ocamlc.opt no longer tries to dynamically load bytecode shared libraries.

Right.  You can close this PR.

**Comment by dra27** [comment]
@xavierleroy - the PR is still a good idea, I think? @stedolan was referring to not requiring a hack of mine in https://github.com/ocaml/ocaml/commit/17e18ad13af77119a33dbfb6d3d3a9b4ba4e77bb being necessary to work around `ocamlc.opt` loading bytecode libraries.

**Comment by stedolan** [comment]
Yes, I'd still like to merge this PR to fix the issue in #8807. This PR makes the native and bytecode `caml_do_local_roots` functions have different names, turning a mysterious segfault into a link error for badly-linked programs. My original implementation didn't work before because the bytecode ocamlc.opt compiler used to dynamically load bytecode libraries into a native runtime, but that's fixed by #9551.

**Comment by xavierleroy** [comment]
OK, I was confused about what the original PR does. 

 I'm neutral about this PR.  On the one hand it's a small change. On the other, users who link manually instead of going through ocamlc or ocamlopt are asking for trouble already, and if they pick the wrong runtime library it's their problem.

**Review by dra27 — APPROVED** [review]
The original issue arose in a situation where the main program is in C, so although it's definitely an advanced use-case, I don't think the originator was necessarily being daft (beyond inadvertently using the wrong library, although it's not as if it's called `libthreadsbyte.a`!

**Comment by stedolan** [comment]
> I'm neutral about this PR. On the one hand it's a small change. On the other, users who link manually instead of going through ocamlc or ocamlopt are asking for trouble already, and if they pick the wrong runtime library it's their problem.

I definitely agree that if someone devises their own linking sequence and breaks it, then that's their own problem. However, this particular case is (a) especially nasty: the program compiles and links without errors or warnings, then starts successfully and blows up at the first GC, and (b) especially easy to fix: two different functions should have different names.

**Comment by stedolan** [comment]
I think "approval" + "neutral" adds up to approval, so merging!

**Pull request merged**
